### DISCUSSION TITLE
SQL Server - Statement(s) could not be prepared

### DISCUSSION THREAD

**Discussion by Pieter3033** [sqlpage/SQLPage#102] [Q&A] [answered]
Attempting to connect to Microsoft SQL Server.

I am using the beneath format for my connection string: (I am connecting to a outside server)
`mssql://username:password@ip:port/DatabaseName`

I get this error when running the executable:
```
[2023-10-13T11:13:44Z WARN  sqlx_core_oldapi::pool::connection] error occurred while testing the connection on-release: error returned from database: Statement(s) could not be prepared.
[2023-10-13T11:13:44Z INFO  sqlpage::webserver::database::migrations] Not applying database migrations because 'C:\Users\user\Desktop\SQL Page\SQLPage Content\sqlpage\sqlpage/migrations' does not exist
```

The above is the 2nd and 3rd line, the 1st line indicates it is attempting to connect to my inputted connection string.
The 4th line states it is connecting locally on 0.0.0.0:8080

Note: I did create a file named sqlpage.json in a folder called sqlpage inside of my websites root folder.

Thanks in advance.

**Comment by lovasoa** [comment, marked as answer]
Hello and welcome to SQLPage !

  The log line `error returned from database: Statement(s) could not be prepared.` seems to indicate that you actually connected successfully to the database. 
   After connecting successfully, SQLPage sent a simple test sql statement comprising of a single comment (`/* SQLx ping */`) to the database, and **that** failed, because the database returned an error saying that the statement could not be prepared. This is just a warning, though, not an error. Does this prevent you from running any sql file through SQLPage ? What do the database logs say ?

**Reply by Pieter3033** [reply]
My sql file content was writing the wrong query, my fault.
Works perfectly now.

Thank you for your quick response, helps a lot.

**Reply by lovasoa** [reply]
can you share the version that didn't work ? Maybe we can improve the error message...

**Reply by Pieter3033** [reply]
The original one was only syntax errors:

SELECT Name FROM Names_Test (Original)

SELECT Name FROM Name_Test (Error generated)

I'm a newbie with SQL and the coding community, I only understood the error message after you explained it and my Senior showed me.
Though I understand now "error returned from database: Statement(s) could not be prepared" means that the database does not recognize the SQL statement that was provided.

**Reply by lovasoa** [reply]
The error message `error returned from database: Statement(s) could not be prepared` is far from ideal. Error messages should already have improved significantly in version 0.12 and then again in 0.13. I'll see if there is something I can do for mssql specifically. 

SQLPage wants to be appealing to people with different backgrounds and not only to professional programmers, so making the  debugging experience painless is an explicit goal.

**Reply by lovasoa** [reply]
v0.14 will have line numbers in sql server error messages

**Reply by Pieter3033** [reply]
I believe everyone appreciates the fact that you optimize SQLPage for everyone, doing a real good job on this one.

I'll let you know if I don't understand something to assist with the optimization for newbies.
